rpki-validator-app/src/main/scala/net/ripe/rpki/validator/models/validation/ValidationProcess.scala

```diff
@@ -110,60 +110,59 @@ class TrustAnchorValidationProcess(override val trustAnchorLocator: TrustAnchorL
-
-    val fetchErrorChecks = fetchErrors.map(e =>
-      new ValidationCheck(ValidationStatus.FETCH_ERROR, ValidationString.VALIDATOR_REPOSITORY_OBJECT_NOT_FOUND, e.url.toString, e.message))
-
-    val certificates = store.getCertificates(taCertUri.toString)
-    if (certificates.size > 1) {
-      validationResult.warnForLocation(
-        new ValidationLocation(taCertUri),
-        ValidationString.VALIDATOR_REPOSITORY_TA_CERT_URI_NOT_UNIQUE,
-        taCertUri.toString)
-    }
-    val matchingCertificates = certificates.filter(keyInfoMatches)
-    if (matchingCertificates.size == 1) {
-      validationResult.rejectIfFalse(keyInfoMatches(matchingCertificates.head), ValidationString.TRUST_ANCHOR_PUBLIC_KEY_MATCH)
-      store.updateValidationTimestamp(Seq(matchingCertificates.head.hash), validationStart)
-    } else {
-      if (matchingCertificates.size > 1) {
-        validationResult.rejectForLocation(new ValidationLocation(taCertUri),
-          ValidationString.VALIDATOR_REPOSITORY_TA_CERT_NOT_UNIQUE,
-          taCertUri.toString)
-      } else {
-        validationResult.rejectForLocation(new ValidationLocation(taCertUri),
-          ValidationString.VALIDATOR_REPOSITORY_OBJECT_NOT_IN_CACHE,
-          taCertUri.toString,
-          "n/a")
+    val firstTaCertUri = certificateLocations.head
+    val validationResult = ValidationResult.withLocation(firstTaCertUri)
+
+    val validCertificateObjects = certificateLocations.flatMap { uri =>
+      RepoService.locker.locked(uri) {
+        repoService.visitTrustAnchorCertificate(uri) match {
+          case Left(errors) =>
+            fetchErrors.appendAll(errors)
+            Seq()
+          case Right(certificateObject) =>
+            if (keyInfoMatches(certificateObject.decoded)) {
+              store.delete(uri)
+              store.storeCertificate(certificateObject)
+              RepoServiceStore.updateLastFetchTime(uri, validationStart)
+              store.updateValidationTimestamp(Seq(certificateObject.hash), validationStart)
+              Seq(certificateObject)
+            } else {
+              validationResult.rejectForLocation(new ValidationLocation(uri), ValidationString.TRUST_ANCHOR_PUBLIC_KEY_MATCH)
+              Seq()
+            }
+        }
       }
     }
 
-    if (validationResult.hasFailureForCurrentLocation)
-      ValidatedObject.invalid(None, Lists.newArrayList("No trust anchor certificate"), taCertUri, None,
-        (validationResult.getAllValidationChecksForCurrentLocation.asScala ++ fetchErrorChecks).toSet)
-    else {
-      val taCertificate = matchingCertificates.head
-      ValidatedObject.valid(
-        Some("cert", taCertificate),
-        Lists.newArrayList(taCertificate.decoded.getSubject.getName),
-        taCertUri,
-        Some(taCertificate.hash),
-        (validationResult.getAllValidationChecksForCurrentLocation.asScala ++ fetchErrorChecks).toSet,
-        taCertificate.decoded
-      )
+    validCertificateObjects.headOption match {
```

Comment:
Should we simply take the headOption or have the check with 
ValidationString.VALIDATOR_REPOSITORY_TA_CERT_URI_NOT_UNIQUE ?

Reply:
Yes, I'll submit an update with headOption map ... getOrElse ...
The VALIDATOR_REPOSITORY_TA_CERT_URI_NOT_UNIQUE should not be possible anymore with the new logic (with old logic it was possible to have objects with the same URI but different hashes _in the store_, but we do not use store at this point anymore).